### PR TITLE
Modem trace shell send uart

### PR DESCRIPTION
**Pull request by gregersrygg**
Add shell command to send stored modem traces over UART when the device tree chosen node "nordic,modem-trace-uart" is specified.

This PR extends #11095 with a command to send stored modem traces to UART. This is useful when testing storage of modem traces and can be used for testing.

The chosen node name `nordic,modem-trace-uart` is the same as used in #10891. So that PR should get merged first.

### CODE REVIEW

File: samples/nrf9160/modem_trace_flash/src/dump_uart.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef MODEM_TRACE_FLASH_DUMP_UART_H
+#define MODEM_TRACE_FLASH_DUMP_UART_H
+
+static void print_traces(void);
```
Comment (balaji-nordic) [inline samples/nrf9160/modem_trace_flash/src/dump_uart.h:10]:
```suggestion
void print_traces(void);
```

File: samples/nrf9160/modem_trace_flash/src/dump_uart.c [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	if (uart_dev == NULL || !device_is_ready(uart_dev)) {
+		LOG_ERR("uart1 device not found/ready!");
+		return;
+	}
+	for (int i = 0; i < len; i++) {
+		uart_poll_out(uart_dev, buf[i]);
+	}
+}
+
+static void print_traces(void)
```
Comment (balaji-nordic) [inline samples/nrf9160/modem_trace_flash/src/dump_uart.c:28]:
```suggestion
void print_traces(void)
```

File: samples/nrf9160/modem_trace_flash/src/main.c [hunk 3]
```diff
@@ -33,45 +25,15 @@ static void on_cfun(enum lte_lc_func_mode mode, void *context)
...
-{
-	if (!device_is_ready(uart_dev)) {
-		LOG_ERR("uart1 device not found/ready!");
-		return;
-	}
-	for (int i = 0; i < len; i++) {
-		uart_poll_out(uart_dev, buf[i]);
-	}
-}
+
```
Comment (balaji-nordic) [inline samples/nrf9160/modem_trace_flash/src/main.c:28]: Spurious new line?

File: samples/nrf9160/modem_trace_flash/src/main.c [hunk 4]
```diff
@@ -33,45 +25,15 @@ static void on_cfun(enum lte_lc_func_mode mode, void *context)
...
-		}
-		if (ret == 0) {
-			LOG_DBG("No more traces to read from flash");
-			break;
-		}
-
-		read_offset += ret;
-		print_uart1(read_buf, ret);
+	if (IS_ENABLED(MODEM_TRACE_FLASH_SAMPLE_UART)) {
+		print_traces();
```
Comment (balaji-nordic) [inline samples/nrf9160/modem_trace_flash/src/main.c:33]: Please use a different name for this function. Now it looks like it calls itself.

**Review by balaji-nordic — CHANGES_REQUESTED** [review]

File: doc/nrf/libraries/modem/nrf_modem_lib.rst [hunk 5]
```diff
@@ -329,6 +329,76 @@ Other configurations set in the devicetree, such as the current speed, are overw
...
+* Delete all stored modem traces:
+
+  .. code-block:: console
+
+     modem_trace clear
+
+Send to Memfault
+----------------
+
+To register an account and obtain the project key, refer to the :ref:`using_memfault` section of the :ref:`ug_memfault` guide.
```
[the remark this answers is not in the document]
Comment (gregersrygg) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:367]: @divipillai ?

File: doc/nrf/libraries/modem/nrf_modem_lib.rst [hunk 6]
```diff
@@ -329,6 +329,76 @@ Other configurations set in the devicetree, such as the current speed, are overw
...
+.. figure:: /images/modem_shell_trace_download.png
+   :alt: Modem trace download
+
+   Modem trace download
+
+Example
+--------
+
+The following is an example of stored modem traces sent to Memfault.
+This will free up the stored traces as they are sent.
```
[the remark this answers is not in the document]
Comment (gregersrygg) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:394]: As they are sent

File: doc/nrf/releases/release-notes-changelog.rst [hunk 7]
```diff
@@ -333,6 +333,12 @@ Modem libraries
 
   * Updated the library to allow a ``PDP_type``-only configuration in the :c:func:`pdn_ctx_configure` function.
 
+* :ref:`nrf_modem_lib_readme` library:
+
+  * Added:
+
+    * Moved modem_trace shell command from modem_shell sample into nrf_modem_lib to be used in any app with :kconfig:option:`CONFIG_SHELL` and :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE` enabled.
```
[the remark this answers is not in the document]
Comment (gregersrygg) [reply doc/nrf/releases/release-notes-changelog.rst:340]: Ah, the joy of PR dependencies 😞 This change is from #11095. You should only look at the last commit when reviewing this PR.

File: lib/nrf_modem_lib/shell/modem_trace_memfault.h [hunk 8]
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef SHELL_MODEM_TRACE_MEMFAULT_H
+#define SHELL_MODEM_TRACE_MEMFAULT_H
```
[the remark this answers is not in the document]
Comment (gregersrygg) [reply lib/nrf_modem_lib/shell/modem_trace_memfault.h:8]: This change belongs to #11095.

File: lib/nrf_modem_lib/shell/modem_trace_memfault.h [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+#include <stddef.h>
+#include <stdint.h>
+#include <zephyr/shell/shell.h>
+
+void modem_trace_memfault_send(const struct shell *sh, size_t size, uint32_t trace_duration_ms);
+
+#endif /* SHELL_MODEM_TRACE_MEMFAULT_H */
```
[the remark this answers is not in the document]
Comment (gregersrygg) [reply lib/nrf_modem_lib/shell/modem_trace_memfault.h:16]: This change belongs to #11095.

File: lib/nrf_modem_lib/shell/modem_trace_shell.c [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+	int err;
+
+	shell_print(sh, "Clearing modem traces");
+
+	err = nrf_modem_lib_trace_clear();
+	if (err == -ENOTSUP) {
+		shell_error(sh, "The current modem trace backend does not support this operation");
+	} else if (err) {
+		shell_error(sh, "Failed to clear modem traces: %d", err);
+	}
```
[the remark this answers is not in the document]
Comment (gregersrygg) [reply lib/nrf_modem_lib/shell/modem_trace_shell.c:76]: This change belongs to #11095.

File: doc/nrf/libraries/modem/nrf_modem_lib.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (divipillai) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:367]: In PR https://github.com/nrfconnect/sdk-nrf/pull/11095/files, the docs were updates as : 

```suggestion
To send data to Memfault, you must have a `Memfault account <Memfault registration page_>`_, a project key, and :ref:`configure Memfault <ug_memfault_config>` in your application.
Refer to the :ref:`ug_memfault` guide for more information.
```

Please note the details about  remote debugging is mentioned in the Memfault user guide. May be we could change this as -

To send stored modem traces to Memfault, you must........

File: doc/nrf/libraries/modem/nrf_modem_lib.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (divipillai) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:394]: Should be "after they are sent". I overlooked it. Thanks for pointing this out.

File: doc/nrf/libraries/modem/nrf_modem_lib.rst [hunk 11]
```diff
@@ -329,6 +329,76 @@ Other configurations set in the devicetree, such as the current speed, are overw
...
    When one of the UART trace backends is enabled by either the Kconfig option :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_UART` or :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_UART_SYNC`, it initializes the UART1 driver, regardless of its status in the devicetree.
 
+Modem trace shell command
+==========================
+
+Shell command: ``modem_trace``
+
+Use the modem trace commands to control the trace functionality in the modem when :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_SHELL` is enabled.
+
+A trace backend that can store modem traces is required to upload modem traces to the cloud.
```
[the remark this answers is not in the document]
Comment (balaji-nordic) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:339]: okay. Fine by me. :)

File: doc/nrf/releases/release-notes-changelog.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (balaji-nordic) [reply doc/nrf/releases/release-notes-changelog.rst:340]: ah ok

File: lib/nrf_modem_lib/shell/Kconfig [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+	depends on SHELL
+	depends on NRF_MODEM_LIB_TRACE
+
+if NRF_MODEM_LIB_TRACE_SHELL
+
+DT_CHOSEN_NORDIC_MODEM_TRACE_UART := nordic,modem-trace-uart
+config NRF_MODEM_LIB_TRACE_SHELL_UART
+	bool
+	default $(dt_chosen_enabled,$(DT_CHOSEN_NORDIC_MODEM_TRACE_UART))
+
```
[the remark this answers is not in the document]
Comment (balaji-nordic) [reply lib/nrf_modem_lib/shell/Kconfig:13]: ah okay. this is fine then

File: lib/nrf_modem_lib/shell/modem_trace_shell.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (balaji-nordic) [reply lib/nrf_modem_lib/shell/modem_trace_shell.c:76]: okay. commented in that PR. Will resolve this.

File: doc/nrf/libraries/modem/nrf_modem_lib.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (divipillai) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:339]: Addressed in PR https://github.com/nrfconnect/sdk-nrf/pull/11095

File: doc/nrf/libraries/modem/nrf_modem_lib.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (divipillai) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:367]: Addressed in PR https://github.com/nrfconnect/sdk-nrf/pull/11095

File: doc/nrf/libraries/modem/nrf_modem_lib.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (divipillai) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:394]: Addressed in PR https://github.com/nrfconnect/sdk-nrf/pull/11095

**Pull request closed without merging**